### PR TITLE
fix(parse): retain metadata-only diff files

### LINKED ISSUE #3250: fix(mcp): preserve binary, rename, and whitespace-only diff metadata
## Problem

`parseDiffHunks` currently only recognizes `+++ b/...` headers and `@@ ... @@` hunks. Binary-only diffs, rename-only diffs (including similarity/mode metadata), and whitespace-only changes can therefore be dropped or represented without stable file metadata before `detect_changes` builds its impact map.

This is related to the recent `detect_changes` work in #2915 / [PR #2930](https://github.com/abhigyanpatwari/GitNexus/pull/2930), and the edge cases described in #2838 and #3032, but is a separate parser/test gap.

## Proposed scope

- keep the existing line-oriented parser and API; add explicit metadata handling for binary, rename, similarity-index, and mode-only records;
- keep the existing `--no-color` / `--default-prefix` call-site guarantees;
- add focused unit cases (and a bounded fuzz/property-style corpus if compatible) for binary, rename, whitespace-only, and mixed metadata diffs;
- avoid changes to graph traversal or Cypher generation.

The expected first PR should be limited to `gitnexus/src/storage/git.ts` and its parser tests, with a stable `FileDiff` result for callers that choose to skip non-symbol-bearing changes.

### PR DESCRIPTION
**Pull request by auyua9**
## Summary

- retain file entries from `diff --git` headers for binary, rename-only, and mode-only changes
- keep line-oriented parsing and avoid inventing symbol hunks for metadata-only files
- add focused parser coverage for binary, rename-only, and whitespace-only diffs

Closes #3250

## Validation

- `git diff --check` passed
- Vitest not run: this fresh worktree has no installed dependencies (`node_modules/.bin/vitest` absent)

---

## Summary

*This appears to be a metadata-only diff parsing change spanning Git storage parsing, local MCP change detection, and focused unit coverage. Its reach is broad through connected CLI, server, and integration paths, though the changed files carry LOW file risk.*

**🔴 CRITICAL blast radius. A Git diff parsing and local change detection update in `gitnexus/src/storage/git.ts` and `gitnexus/src/mcp/local/local-backend.ts` reaches 18 total dependents.**

Review `parseDiffHunks`, `coalesceHunks`, `coalesceHunksByPath`, `hunksOverlapRange`, and `GraphLineRange` first in `gitnexus/src/storage/git.ts`, which is the hottest changed file. Then verify how `buildDetectChangesDiffArgs`, `LocalBackend`, and `detectChanges` in `gitnexus/src/mcp/local/local-backend.ts` consume the resulting diff and hunk data.

Impact lands across `Cli`, `Local`, `Storage`, `Mcp`, `Server`, and `Integration`, with 29 affected flows. The accompanying coverage is concentrated in `gitnexus/test/unit/detect-changes-eol.test.ts`, `gitnexus/test/unit/detect-changes-hunk-scale.test.ts`, and `gitnexus/test/unit/parse-diff-hunks.test.ts`, so reviewers should check that metadata-only files remain represented consistently through parsing, hunk coalescing, and detection.

_Added by GitNexus for PR #3251. Edit freely — this block is replaced on the next review, everything above it is left untouched._

### CODE REVIEW

**Review by koriyoshi2041 — COMMENTED** [review]
I reproduced a metadata-only case that the new header parser still drops: Git C-quotes paths containing non-ASCII/control bytes under the default `core.quotePath` behavior.

```ts
parseDiffHunks([
  'diff --git "a/assets/\\344\\270\\255.png" "b/assets/\\344\\270\\255.png"',
  'Binary files differ',
].join('\n'))
// []
```

There is a second ambiguity when an unquoted path itself contains ` b/`: `diff --git a/dir b/name.png b/dir b/name.png` currently returns `name.png` rather than `dir b/name.png` because both captures are greedy.

The focused suite passes at `b10fd92e` (12/12), as does `tsc --noEmit`. Could we parse/decode Git's quoted header tokens (or cover these two cases with another unambiguous source of the destination path) before relying on the header for metadata-only files?

File: gitnexus/src/storage/git.ts [hunk 1]
```diff
@@ -772,9 +782,18 @@ export function parseDiffHunks(diffOutput: string): FileDiff[] {
   const files: FileDiff[] = [];
   let current: FileDiff | null = null;
   for (const line of diffOutput.split('\n')) {
-    if (line.startsWith('+++ b/')) {
-      current = { filePath: line.slice(6), hunks: [] };
-      files.push(current);
+    if (line.startsWith('diff --git ')) {
```
Comment (magyargergo) [inline gitnexus/src/storage/git.ts]: **P1** Sticky `current`: an unparsed `diff --git` does not clear the previous file, so a later C-quoted file’s `@@` hunks attach here and skip the parse-failure arm.

Trigger: mode-only (or binary / 100% rename) `script.sh`, then an edit to a default-`core.quotePath` path such as `你好.ts`. Git emits unquoted `diff --git a/script.sh b/script.sh` then `diff --git "a/src/\\344\\275\\240\\345\\245\\275.ts" "b/…"`. `filePathFromGitHeader` requires unquoted `a/… b/…`, so this `if (filePath)` does not run and `current` stays `script.sh`. Quoted `+++ "b/…"` does not match `+++ b/`. The hunk is recorded on `script.sh`; the CJK path is absent.

[reproduced] HEAD: `[{ filePath: "script.sh", hunks: [{ startLine: 1, endLine: 1 }] }]`. BASE on the same bytes: `[]` → `local-backend.ts:5611` `unknown`+`partial`, CLI `tool.ts:90-95` exit 1. After this PR, `fileDiffs.length > 0` disables that arm and the gate exits 0.

Isolated C-quoted metadata-only diffs still return `[]` (fail-closed) — `koriyoshi2041` already flagged that drop. The *new* honesty hole is this cascade.

Fix: set `current = null` on every `diff --git` line first; count unparsed headers; `detectChanges` must set `partial: true` and `risk_level: 'unknown'` when any header failed. Parse C-quoted tokens or pass `-c core.quotePath=false`.

File: gitnexus/src/storage/git.ts [hunk 2]
```diff
@@ -772,9 +782,18 @@ export function parseDiffHunks(diffOutput: string): FileDiff[] {
...
+      const filePath = filePathFromGitHeader(line);
+      if (filePath) {
+        current = { filePath, hunks: [] };
+        files.push(current);
+      }
+    } else if (line.startsWith('+++ b/')) {
+      const filePath = line.slice(6);
```
Comment (magyargergo) [inline gitnexus/src/storage/git.ts]: **P1** `line.slice(6)` keeps git’s terminating TAB on `+++`, so a spaced path fails the new equality guard and becomes two `FileDiff`s.

Trigger: edit a tracked file whose path contains a space (`My Documents/file.ts`). Git does not C-quote spaces; it appends a TAB on `---` / `+++` and not on `diff --git`. Header dest is `"My Documents/file.ts"`; this slice is `"My Documents/file.ts\t"`.

[reproduced] HEAD: `[{ filePath: "My Documents/file.ts", hunks: [] }, { filePath: "My Documents/file.ts\t", hunks: [{ startLine: 1, endLine: 1 }] }]`. `changed_files` (`local-backend.ts:5888` Set) is 2 for one file. The tabbed-key symbol miss vs the index is pre-existing on BASE; the new damage is the duplicate + inflated count.

Fix: parse `+++ b/<path>` per unified-diff grammar (path ends at the first TAB) before compare/store.

Related: the same-path skip is otherwise correct. `detect-changes-hunk-scale.test.ts:368` still expects `['code.py','code.py']` and fails on HEAD (`['code.py']`). Update that assertion; do not restore the phantom second entry.

File: gitnexus/src/storage/git.ts [hunk 3]
```diff
@@ -750,10 +750,20 @@ export interface FileDiff {
   hunks: DiffHunk[];
 }
 
+function filePathFromGitHeader(line: string): string | undefined {
+  const match = line.match(/^diff --git a\/(.+) b\/(.+)$/);
```
Comment (magyargergo) [inline gitnexus/src/storage/git.ts]: **P2** Greedy `a/(.+) b/(.+)` takes the last ` b/`, so a dest containing that substring is the wrong filename — and this PR now publishes that name for metadata-only files.

Trigger: binary / mode-only (no `+++ b/` to correct) `diff --git a/foo b/bar.png b/foo b/bar.png`. Dest becomes `bar.png`. A 100% rename `a/plain.ts b/foo b/plain.ts` yields the source (`plain.ts`), not the dest.

[reproduced] Coordinator + Claude correctness/adversarial/risk. `koriyoshi2041` already reported this on the PR; it still holds on `b10fd92e`. BASE returned `[]` (parse-failure). Spaces *without* ` b/` (`My Documents/file.ts`) parse; `/b/` without a preceding space (`a/b/foo.ts`) also parses.

Fix: when src==dst, recover `name` from `a/${name} b/${name}`. For renames prefer `rename to` / a tab-stripped `+++` over the greedy split.

**Review by magyargergo — COMMENTED** [review]
# PR #3251 tri-review digest (`fix(parse): retain metadata-only diff files`)

HEAD `b10fd92e` · BASE `506432017` · worktree `/workspace/.worktrees/pr-3251`

**Branch hygiene:** `clean feature/fix PR`  
**Merge state:** `mergeable`  
**Final verdict:** `not production-ready`

## Engine breakdown + Codex liveness

| Engine | Legs | Status |
|---|---|---|
| **A — Claude** | Swarm (facts, hygiene, risk, test-ci, security, docs) + CE (correctness, adversarial, testing, security, project-standards, maintainability) | Live. Findings below are from these lanes only. |
| **B — Codex (GPT-5.5)** | B1 swarm · B2 ce · B3 adversarial | **All three failed** (workspace spend cap). Artifacts are error markers, not reviews. B3 also failed structured-JSON parse for the same reason. |

**Codex unavailable — both-engine review degraded to Claude-only.** There is **no** cross-engine corroboration anywhere in this digest. Where several Claude lanes agree, that is **consistent within one engine** (shared priors / shared blind spots), not independent confirmation.

---

## 1. Review bar for this PR

`detect_changes` must not look complete when it missed or mis-attributed an edit (pre-commit honesty: #2915 / #2283 / #3131). Metadata-only paths may be retained as `FileDiff`s without inventing symbol ranges (#3250). C-quoted `core.quotePath` headers, unified-diff TABs on `+++`, and dests containing ` b/` must not silently remap hunks or publish the wrong path. Tests that lock parse shape must match HEAD. Fork CI / Vercel deploy-auth are not substitutes for a green unit suite.

## 2. Problem being solved

#3250: `parseDiffHunks` only recognized `+++ b/` + `@@` hunks, so binary-only, rename-only, and mode-only diffs were dropped before `detect_changes` built its impact map. This PR keeps the line-oriented parser and opens a `FileDiff` from `diff --git a/X b/Y` (destination Y) so those files are retained with empty hunks.

## 3. Current PR state

OPEN, not draft. Fork PR from `auyua9/gitnexus` branch `codex/fix-diff-metadata` into `abhigyanpatwari/GitNexus` `main`. HEAD `b10fd92e7bcebf45f3cf07e201f63de7ea779132`. One linear commit (`fix(parse): retain metadata-only diff files`). Two files: `gitnexus/src/storage/git.ts`, `gitnexus/test/unit/parse-diff-hunks.test.ts` (+55/−3). `Closes #3250`. One existing COMMENT review from `koriyoshi2041` (no inlines). No prior `<!-- pr-tri-review -->`.

## 4. Merge status and mergeability

**Merge state:** `mergeable`

GitHub REST: `mergeable: true`, `mergeable_state: behind`, `rebaseable: true`, not draft. Local `git merge-tree origin/main HEAD` is clean. Behind current `main` by 9 commits that do **not** touch the PR paths — `behind` is evidence, not the classification. Vercel FAILURE is deploy authorization, not a product check; do not classify as `checks failing`. Fork `pull_request` workflows are `action_required` (visibility gap for tests, not a GitHub merge conflict).

## 5. Repository history considered

- #3250 (OPEN): first-PR scope is parser + tests; stable `FileDiff` for callers that skip non-symbol-bearing changes. Related: #2915 / PR #2930 (hunk scale), #2838 / #2839 (EOL), #3131 / #3138 (color / zero-symbol honesty). #3032 is citation-only.
- `koriyoshi2041` already COMMENT-reviewed this SHA: C-quoted `core.quotePath` headers drop metadata-only files; greedy ` b/` dest splits. This digest credits that review. The **new** honesty hole is the sticky-`current` cascade.
- Prior session memory (#50837 / #50916): `noprefix` / `mnemonicPrefix` false all-clears — call-site prefix pins are unchanged and still required.

## 6. Branch hygiene assessment

**Branch hygiene:** `clean feature/fix PR`

One commit, parent exactly BASE `506432017`. No merge-from-main. Diff is parser + matching unit tests only. Nine later `main` commits do not overlap these paths. `merge-tree` clean. Do **not** emit `rebase/split required`. Optional update-from-main before merge is polish.

## 7. Understanding of the change

`parseDiffHunks` now opens a `FileDiff` on `diff --git a/X b/Y` via `filePathFromGitHeader` (`/^diff --git a\/(.+) b\/(.+)$/`, dest = group 2). `+++ b/` only pushes a new entry when `line.slice(6) !== current.filePath`. Empty `hunks` are intentional. `detectChanges` and CLI `output()` are unchanged: `fileDiffs.length === 0` is still the only parse-failure detector; CLI exits 1 only on `error` or `partial === true`. `coalesceHunksByPath` still drops empty-hunk paths from query bounds; `changed_files` is a Set of `fileDiff.filePath`.

## 8. Findings

### What is solid

The advertised parser half of #3250 works for the fixtures it tests:

- Unquoted ASCII `diff --git a/X b/X` **does** retain binary-only and 100% rename-only files as `{ hunks: [] }`.
- The full `parse-diff-hunks.test.ts` file is **12/12 pass** on HEAD (9 pre-existing + 3 new). Binary and rename-dest cases lock real new behavior.
- Same-path phantom `+++ b/<self>` under `-U0` is **no longer** a second `FileDiff` — a production improvement over BASE.
- Real `+++ /dev/null` deletions now keep the git-header path and the `+0,0` hunk (BASE returned `[]`). That is a fix, not a hole.
- Isolated C-quoted / non-ASCII diffs still parse to `[]` → `unknown` + `partial` + CLI exit 1 (fail-closed). Prefix pinning (`--src-prefix=a/ --dst-prefix=b/ --color=never --no-ext-diff`) is intact.
- `coalesceHunksByPath` still skips empty `hunks` (`git.ts:887`); `hunks[0]` / Ladybug empty-bounds crash is **refuted**. ReDoS, path-escape-to-FS, and Cypher/shell injection were probed and **refuted** (no sink).

### F1 — Sticky `current` / C-quote cascade

- **Risk:** A parsed metadata `diff --git` leaves `current` live. A later C-quoted file’s `@@` hunks attach to the metadata path; the quoted file is absent. `fileDiffs.length > 0` skips the parse-failure arm. CLI exits 0. BASE on the same bytes was fail-closed (`unknown`+`partial`+exit 1). Isolated quoted drop already flagged by `koriyoshi2041`; the *new* hole is this cascade.
- **Evidence to check:** `git.ts:785–790` (update `current` only inside `if (filePath)`). Trigger: mode-only `script.sh` then C-quoted `你好.ts`. [reproduced] HEAD: `[{ filePath: "script.sh", hunks: [{ startLine: 1, endLine: 1 }] }]`. `local-backend.ts:5611`; `tool.ts:90-95`. Claude: CE adversarial F1, swarm risk F1.
- **Recommended fix:** Set `current = null` on every `diff --git` first; count unparsed headers; `detectChanges` must set `partial: true` and `risk_level: 'unknown'` when any header failed. Parse C-quoted tokens or pass `-c core.quotePath=false`.
- **Blocks merge:** yes

### F2 — Spaced-path `+++` TAB split

- **Risk:** Git appends a TAB on `+++` when the path contains a space (not on `diff --git`). `slice(6)` keeps the tab; the new equality guard opens two `FileDiff`s. `changed_files` becomes 2. Tabbed-key symbol miss vs the index is pre-existing on BASE; new damage is the duplicate + inflated count.
- **Evidence to check:** `git.ts:792–796`. Trigger: `My Documents/file.ts`. [reproduced] HEAD: clean empty + tabbed-with-hunks. `local-backend.ts:5888` Set. Claude: CE correctness F1 (P1), CE adversarial / swarm risk (P2). Highest severity kept: P1.
- **Recommended fix:** Parse `+++ b/<path>` per unified-diff grammar (path ends at the first TAB) before compare/store.
- **Blocks merge:** yes

### F3 — Greedy ` b/` dest

- **Risk:** `a/(.+) b/(.+)` takes the last ` b/`. This PR now **publishes** that wrong dest for metadata-only files (no `+++` rescue). BASE returned `[]` (parse-failure). Credit `koriyoshi2041`.
- **Evidence to check:** `git.ts:754`. Trigger: `diff --git a/foo b/bar.png b/foo b/bar.png` → `bar.png`. [reproduced]. Claude: CE correctness F2, CE adversarial F3-B, swarm risk F3-B.
- **Recommended fix:** When src==dst, recover `name` from `a/${name} b/${name}`. For renames prefer `rename to` / a tab-stripped `+++` over the greedy split.
- **Blocks merge:** yes

### F4 — Hunk-scale assertion stale

- **Risk:** `detect-changes-hunk-scale.test.ts:368` still expects `['code.py','code.py']` for a same-path `++ b/code.py` content line. HEAD correctly emits `['code.py']`. Vitest **FAIL**. The skip is intended; the test is not. Ubuntu CI will fail this file if the fork workflow is approved.
- **Evidence to check:** Test 352–372; cause `git.ts:791–796`. Coordinator Vitest: `expected [ 'code.py' ] to deeply equal [ 'code.py', 'code.py' ]`. Claude test-ci P0; CE testing P1. Treated as **P1 test hygiene**.
- **Recommended fix:** Update the assertion to `['code.py']`. Do not restore the phantom second entry. Add a parser unit case for `diff --git` + header `+++ b/` + content `+++ b/` → one `FileDiff`.
- **Blocks merge:** yes

## 9. PR-specific assessment sections

**Parser honesty.** Isolated unquoted ASCII binary/rename retention works. Mixed metadata + C-quoted sibling is a silent-pass vs BASE fail-closed. Spaced TAB and greedy ` b/` publish wrong/duplicate paths.

**Test / CI.** Parser 12/12 pass locally. Hunk-scale “counts a path” **FAIL**. No GitHub-visible unit/integration/typecheck (`action_required` fork workflows). No mode-only fixture (JSDoc claims it). Whitespace new case still has `+++ b/` and would pass on BASE. No quoted / TAB / ` b/` fixtures. No `detectChanges` consumer test for the sentinel flip.

**Security.** ReDoS, path-escape-to-FS, Cypher/shell injection **refuted** (no sink; production input is pinned git stdout). Metadata-only `risk_level:'low'` is **not** a #2915 `none` all-clear (formatter already has the no-overlap sentence; gate keys on `partial`/`error`).

## 10. Back-and-forth avoided by verifying

- Coordinator `tsx` on HEAD `parseDiffHunks`: sticky current C-quote steal; spaced-path TAB split; greedy ` b/` dest; same-path phantom de-dupe; cross-path content `+++ b/<other>` still opens a second entry (pre-existing).
- Coordinator Vitest: parser 12/12 pass; hunk-scale “counts a path” FAIL at `:368`.
- CLI `tool.ts:90-95` code-read: exit 1 only on `error` or `partial===true`.
- Codex B1/B2/B3: spend-cap error markers only — Engine B unavailable.
- Refuted by code-read / execution: `hunks[0]` crash; isolated C-quoted silent-pass (still fail-closed); `+++ /dev/null` drop; same-path phantom as production `changed_files` inflation; ReDoS / injection.

## 11. Open questions

**Only** the empty-hunk `risk_level:'low'` product question. Isolated binary/rename/mode diffs flip from `unknown`+`partial`+exit 1 to `changed_files>0`, `changed_count:0`, `risk_level:'low'`, no `partial`, CLI exit 0. CE correctness + both security lanes treat this as the advertised #3250 contract. CE adversarial + swarm risk raise **100% rename** to P1 (index still keys the old path; neither side is queried). Not inlined. Not promoted to a new P1 finding. Worth a follow-up on rename-from query / “entire stdout parsed.”

## 12. Final verdict

**`not production-ready`**

The advertised unquoted-ASCII metadata retention works, and same-path phantom `+++ b/` de-dupe is a real improvement, but HEAD introduces a silent-pass: a parsed metadata header plus a later C-quoted sibling steals hunks and exits 0 where BASE failed closed. The new `+++` equality guard also splits spaced paths on git’s terminating TAB and publishes greedy ` b/` dests that BASE never returned. A shipped #2915 unit test fails on that same-path skip. Those four issues block merge. Update-from-main is optional (conflict-free). Do not merge on labeler + GitNexus-bot green.

---

## Proposed inline comments

Anchored only to added/context lines in `git diff $BASE HEAD` on `git.ts`. `detectChanges` / `tool.ts` / the hunk-scale test are unchanged — referenced from the nearest in-diff line.

| # | Path | RIGHT line | Sev | One-line problem | Evidence | Fix | Tag |
|---|---|---|---|---|---|---|---|
| 1 | `gitnexus/src/storage/git.ts` | 785 | P1 | Unparsed `diff --git` leaves `current` live; later C-quoted `@@` hunks attach to the previous metadata file and skip parse-failure. | Mode-only `script.sh` then C-quoted `你好.ts` → hunks on `script.sh`, CJK path absent. BASE `[]` → `partial`. Isolated quoted drop already flagged by koriyoshi2041. | Null `current` on every `diff --git`; count unparsed headers; set `partial`+`unknown` if any failed. Parse C-quotes or `-c core.quotePath=false`. | [reproduced] |
| 2 | `gitnexus/src/storage/git.ts` | 792 | P1 | `+++ b/` `slice(6)` keeps git’s terminating TAB; spaced path ≠ git-header dest → two FileDiffs. | `My Documents/file.ts`: clean empty + tabbed-with-hunks. `changed_files` 2. | Strip path at first TAB before compare/store. | [reproduced] |
| 3 | `gitnexus/src/storage/git.ts` | 754 | P2 | Greedy `(.+) b/(.+)` dest is the suffix after the last ` b/`. | `diff --git a/foo b/bar.png b/foo b/bar.png` → `bar.png`. koriyoshi2041 already reported. | Pair `a/${name} b/${name}` when src==dst; else `rename to` / tab-stripped `+++`. | [reproduced] |

## Lower-priority / test gaps / refuted

**Test / DoD gaps (body only):** no mode-only fixture (JSDoc claims it); no mixed metadata / metadata-then-content parser test; no `detectChanges` consumer test for the sentinel flip; whitespace-only new case still has `+++ b/` and would pass on BASE (false confidence); no quoted / TAB / ` b/` fixtures; stale `+++ b/`-only comments in `detect-changes-eol.test.ts:68` and `local-backend.ts:5612` / `5882–5887`. PR body has no residual-risk / rollback note. `Closes #3250` is only honest if quoted-path + MCP consumer leftovers are waived or followed up.

**Refuted (do not re-litigate):**

- `hunks[0]` / #50398 empty-bounds scan blowup — empty hunks never enter the bounds map.
- Isolated C-quoted path as a **new** silent-pass — still fail-closed; koriyoshi’s drop stands.
- `+++ /dev/null` deletions dropping the file — HEAD keeps the git-header path.
- Same-path phantom `+++ b/` as production `changed_files` inflation — HEAD fixes it; only the stale test fails.
- Cross-path content `+++ b/<other>` hunk steal — **pre-existing** (HEAD == BASE).
- Spaced-path TAB as a **new** symbol miss vs BASE — BASE already queried the tabbed key.
- ReDoS / `../` path traversal / Cypher or shell injection — no sink; production input is pinned git stdout.
- Metadata-only `risk_level:'low'` as a security / #2915 false all-clear — gate keys on `partial`/`error`, not `risk_level`; formatter does not say “No changes detected.”
- `noprefix` / `mnemonicPrefix` / ANSI all-clears — call-site flags unchanged.
- Isolated mode/binary hiding symbol **text** — no line hunks in isolation.

## CI status

Visible on `b10fd92e` (fork `auyua9/gitnexus` → `abhigyanpatwari/GitNexus`):

- **Vercel** FAILURE — deploy authorization, not product CI.
- **Apply conventional label** SUCCESS; **Validate PR title** SKIPPED. `labels: []` despite the green labeler job — verify `bug` before relying on release-note grouping.
- **GitNexus** (`gitnexus-check`) SUCCESS — File-level CRITICAL blast on `parseDiffHunks` is the bot’s expansion, not a verified caller set. `impact(parseDiffHunks)` LOW / empty callers is an **index gap**; production caller is `detectChanges` (HIGH: CLI / MCP / server).
- Fork `pull_request` workflows (**CI**, Gitleaks, CodeQL, Dependency Review, PR Description Check, Docker, Autofix) are `action_required` with empty job lists. **No GitHub-visible unit/integration/typecheck** on this SHA.
- Author stated Vitest was not run. Local targeted run: parser 12/12 pass; hunk-scale “counts a path” **FAIL**.
- PR: OPEN, not draft, `MERGEABLE`, `mergeStateStatus: BEHIND`.

Do not merge on labeler + bot green.

## Coverage limits

- **Claude-only.** Codex B1/B2/B3 produced spend-cap errors only. Do not treat multi-lane Claude agreement as two-engine confirmation.
- Diff is two files (+55/−3). Parser + new tests were read in full. `detectChanges` / CLI gate / formatter were code-read as unchanged consumers, not re-executed end-to-end against a live Ladybug index.
- Graph MCP on this workspace was stale / storage-version mismatched in the facts lane; caller set confirmed by grep, not a fresh `impact`.
- Reproductions used HEAD `parseDiffHunks` via `tsx` / targeted Vitest on coordinator fixtures, not `npm run build`.
- `origin/main` has moved 9 commits since BASE; those commits do not touch the PR paths. Caller line numbers in `local-backend.ts` may still have drifted on `main`.

## Close

Automated multi-tool digest from Claude swarm + Claude CE only. **Codex unavailable — both-engine review degraded to Claude-only.** Verify against the worktree before acting; a wrong maintainer-signed inline is expensive. Highest-confidence posts are the three [reproduced] parser defects above (cascade, TAB split, greedy ` b/`). Update `detect-changes-hunk-scale.test.ts:368` in the same change. Isolated quoted-path drops and greedy ` b/` were first raised by `koriyoshi2041` — the cascade is the new honesty hole.

File: gitnexus/src/storage/git.ts (hunk 1 above)
Comment (magyargergo) [reply gitnexus/src/storage/git.ts]: > an unparsed `diff --git` does not clear the previous file, so a later C-quoted file’s `@@` hunks attach here

Addressed in bba03c486. Every `diff --git` clears `current` first; C-quoted tokens decode as UTF-8; unparsed headers increment a count and `detectChanges` sets `partial` plus `risk_level: 'unknown'`. Production git is also pinned with `-c core.quotePath=false`.

File: gitnexus/src/storage/git.ts (hunk 2 above)
Comment (magyargergo) [reply gitnexus/src/storage/git.ts]: > `line.slice(6)` keeps git’s terminating TAB on `+++`, so a spaced path fails the new equality guard and becomes two `FileDiff`s.

Addressed in bba03c486. `+++` paths are cut at the first TAB before compare/store. The hunk-scale assertion now expects `['code.py']` — the same-path skip stays; the phantom second entry does not.

File: gitnexus/src/storage/git.ts (hunk 3 above)
Comment (magyargergo) [reply gitnexus/src/storage/git.ts]: > Greedy `a/(.+) b/(.+)` takes the last ` b/`, so a dest containing that substring is the wrong filename

Addressed in bba03c486. Same-path headers recover `name` from `a/${name} b/${name}`; renames prefer `rename to` (and a tab-stripped `+++`) over the greedy split.

**Comment by magyargergo** [comment]
> I reproduced a metadata-only case that the new header parser still drops: Git C-quotes paths containing non-ASCII/control bytes under the default `core.quotePath` behavior.

> The advertised unquoted-ASCII metadata retention works … but HEAD introduces a silent-pass: a parsed metadata header plus a later C-quoted sibling steals hunks … The new `+++` equality guard also splits spaced paths on git’s terminating TAB and publishes greedy ` b/` dests … A shipped #2915 unit test fails on that same-path skip.

Addressed in bba03c486. C-quoted `diff --git` / `+++` tokens decode as UTF-8; every `diff --git` clears `current` and an unparsed header fails closed (`partial` + `unknown`); `+++` paths end at the first TAB; same-path dests recover `a/${name} b/${name}` and renames prefer `rename to`. The hunk-scale assertion now expects `['code.py']`. Production git is also pinned with `-c core.quotePath=false`.

File: gitnexus/src/storage/git.ts [hunk 4]
```diff
@@ -769,12 +930,46 @@ export interface FileDiff {
...
+      }
+    } else if (line.startsWith('rename to ')) {
+      const filePath = pathFromRenameTo(line);
+      if (!filePath) continue;
+      if (current) current.filePath = filePath;
+      else {
+        current = { filePath, hunks: [] };
+        files.push(current);
+      }
+    } else if (line.startsWith('+++ ')) {
```
[the remark this answers is not in the document]
Comment (magyargergo) [reply gitnexus/src/storage/git.ts]: > An added source line whose text is `++ "b/other.ts"` is emitted by a unified diff as `+++ "b/other.ts"`

Addressed in 2a846d3ca. `+++` is only treated as a file header before the first `@@` of the current file, so hunk-body lines stay attached to that file.

[the remark this answers is not in the document]
Comment (magyargergo) [reply gitnexus/src/storage/git.ts]: > The new broad `+++ ` branch is required for C-quoted headers, but it runs while consuming hunk body lines too.

Addressed in 2a846d3ca. File-header `+++` is ignored once a hunk has started, so a content line like `+++ "b/generated.ts"` no longer opens a second `FileDiff`.

File: gitnexus/src/storage/git.ts [hunk 5]
```diff
@@ -750,10 +750,171 @@ export interface FileDiff {
...
+function filePathFromGitHeader(line: string): string | undefined {
+  if (!line.startsWith(DIFF_GIT_PREFIX)) return undefined;
+  const rest = line.slice(DIFF_GIT_PREFIX.length);
+
+  if (rest.startsWith('"')) {
+    const srcTok = takeCQuotedToken(rest, 0);
+    if (!srcTok) return undefined;
+    let i = srcTok.end;
+    while (rest[i] === ' ') i++;
+    const destTok = takeCQuotedToken(rest, i);
```
[the remark this answers is not in the document]
Comment (magyargergo) [reply gitnexus/src/storage/git.ts]: > a rename from `old name.ts` to `new.ts` has a header of the form `diff --git "a/old name.ts" b/new.ts`

Addressed in 2a846d3ca. Src and dest tokens are parsed independently, so mixed quoted/unquoted `diff --git` headers no longer increment `unparsedGitHeaders`.
